WordPress/wordpress-develop
# `wp_get_global_settings`: add object cache

Trac ticket: https://core.trac.wordpress.org/ticket/57502
Related to https://github.com/WordPress/gutenberg/issues/45171

This PR backports:

- https://github.com/WordPress/gutenberg/pull/45971
- https://github.com/WordPress/gutenberg/pull/45372
- https://github.com/WordPress/gutenberg/pull/46150

Props @felixarntz, @aristath, @ramonjd, @spacedmonkey, @anton-vlasenko, @jorgefilipecosta, @mmtr, @mcsf

## What

This PR adds object cache to the existing `wp_get_global_settings`.

## Why

This PR improves the performance of a request by:

- **40% for themes with `theme.json`** (from 876.05ms to 520.80ms, a total of -355.25ms)
- **10% for themes without `theme.json`** (from 500.26ms to 450.66ms, a total of -49.60ms)

After this PR, the time a theme with `theme.json` takes to process a request is almost the same (70ms slower) than a theme without `theme.json`. See [full dataset](https://docs.google.com/spreadsheets/d/1ulnIA7EDun3vlMrDEcIowT2MwskU1VfScwQ33Ezbows/edit#gid=0).

## How

Adds a cache to the high-level public method that consumers use to query the settings coming from theme.json APIs (WordPress, block, theme, and user data).

## Test

- Make sure automated test pass.
- Test block theme (TwentyTwentyThree)
    - Create new colors as user from the GS sidebar and verify they are shown in the color components in the block sidebar (site editor, post editor).
    - Modify theme.json settings of the theme
- Test classic theme (TwentyTwenty)
    - Verify colors defined by the theme via `add_theme_support( 'editor-color-palette', $palette )` work as expected.
    - Disable custom colors by adding `add_theme_support( 'disable-custom-colors' );` to the `functions.php` of the theme. Verify that users cannot add custom colors.

## Performance analysis

### How performance was measured

I've tested this using the same mechanism I use for all performance work I've been doing, so the impact of each PR can be compared against each other:

- I have loaded the "Hello World!" post as a logged-out user. Conducted the test 9 times and then calculated the median ([see full dataset](https://docs.google.com/spreadsheets/d/1ulnIA7EDun3vlMrDEcIowT2MwskU1VfScwQ33Ezbows/edit#gid=0)).
- Compared this PR vs trunk at `e27c5a38e371e1db85efeea239accf535227ed39` (svn 55005)
- Used TwentyTwentyThree for theme with `theme.json` and TwentyTwenty for a theme with no `theme.json`.

I'm also sharing the raw data I extracted (cachegrind files) so anyone can open them with their tool of choice and inspect the results. Feel free to DM me if you are interested in setting this up for running the test yourself or anything.

- Data for TwentyTwenty:
  - [`cachegrind.out._2022_12_19_hello-world_.twentytwenty.pr.zip`](https://github.com/WordPress/wordpress-develop/files/10261039/cachegrind.out._2022_12_19_hello-world_.twentytwenty.pr.zip)
  - [`cachegrind.out._2022_12_19_hello-world_.twentytwenty.trunk.zip`](https://github.com/WordPress/wordpress-develop/files/10261040/cachegrind.out._2022_12_19_hello-world_.twentytwenty.trunk.zip)
- Data for TwentyTwentyThree:
  - [`cachegrind.out._2022_12_19_hello-world_.twentytwentythree.pr.zip`](https://github.com/WordPress/wordpress-develop/files/10261044/cachegrind.out._2022_12_19_hello-world_.twentytwentythree.pr.zip)
  - [`cachegrind.out._2022_12_19_hello-world_.twentytwentythree.trunk.zip`](https://github.com/WordPress/wordpress-develop/files/10261045/cachegrind.out._2022_12_19_hello-world_.twentytwentythree.trunk.zip)

### How to reproduce the experiment yourself

You don't need to, as I've made available the cachegrind data (see section above). Though you may still want to run this yourself for peer review:

1. Configure docker and enviroment by applying this patch. This tells docker to enable the xdebug profiler, output the performance files by URL, and set `WP_DEBUG` to false:

```diff
diff --git a/.env b/.env
index 63a8169f64..f3a87fbb08 100644
--- a/.env
+++ b/.env
@@ -18,7 +18,7 @@ LOCAL_DIR=src
 LOCAL_PHP=latest
 
 # Whether or not to enable Xdebug.
-LOCAL_PHP_XDEBUG=false
+LOCAL_PHP_XDEBUG=true
 
 ##
 # The Xdebug features to enable.
@@ -31,7 +31,7 @@ LOCAL_PHP_XDEBUG=false
 #
 # For a full list of accepted values, see https://xdebug.org/docs/all_settings#mode.
 ##
-LOCAL_PHP_XDEBUG_MODE=develop,debug
+LOCAL_PHP_XDEBUG_MODE=profile
 
 # Whether or not to enable Memcached.
 LOCAL_PHP_MEMCACHED=false
@@ -54,7 +54,7 @@ LOCAL_DB_TYPE=mysql
 LOCAL_DB_VERSION=5.7
 
 # The debug settings to add to `wp-config.php`.
-LOCAL_WP_DEBUG=true
+LOCAL_WP_DEBUG=false
 LOCAL_WP_DEBUG_LOG=true
 LOCAL_WP_DEBUG_DISPLAY=true
 LOCAL_SCRIPT_DEBUG=true
diff --git a/docker-compose.yml b/docker-compose.yml
index 739c65f4a8..b781c3fe94 100644
--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -39,6 +39,7 @@ services:
     environment:
       - LOCAL_PHP_XDEBUG=${LOCAL_PHP_XDEBUG-false}
       - XDEBUG_MODE=${LOCAL_PHP_XDEBUG_MODE-develop,debug}
+      - XDEBUG_CONFIG=profiler_output_name=cachegrind.out.%R
       - LOCAL_PHP_MEMCACHED=${LOCAL_PHP_MEMCACHED-false}
       - PHP_FPM_UID=${PHP_FPM_UID-1000}
       - PHP_FPM_GID=${PHP_FPM_GID-1000}
```

2. Build the environment: `npm install && npm run build:dev`.
3. Start docker: `npm run env:start && npm run env:install`.
4. Visit the site and load the "Hello World!" post. It should be something along the lines of http://localhost:8889/2022/12/19/hello-world/
5. Take the cachegrind data generated by xdebug out of docker:

```sh
# List and find the latest timestamp of cachegrind.out.* files
docker exec -it `docker ps -f name=wordpress-develop_php -q` ls -lat --full-time /tmp/
# Copy from docker to manipulate with any tool in your computer
docker cp `docker ps -f name=wordpress-develop_php -q`:'/tmp/cachegrind.out._2022_12_19_hello-world_.gz' ~/Desktop/
```
6. Inspect the time it takes with any tool of your choice. I use kcachegrind, but there are alternatives such as qcachegrind, webgrind, etc. Note that kcachegrind shows times aggregated by 10ns, so the total included time of the request in the following screenshot would be `35 0442 02 * 10 = 350 442 020ns`, which amounts to `350ms`.

![Captura de ecrã de 2022-10-28 13-13-26](https://user-images.githubusercontent.com/583546/198574623-d1c02a25-1a63-4bf2-808a-f4b7e2f16cbb.png)

7. Repeat 9 times and calculate the median.
